src/client/clientmap.cpp

```diff
@@ -288,206 +283,289 @@ void ClientMap::updateDrawList()
-	blocks_seen.getChunk(camera_block).getBits(camera_block) = 0x07; // mark all sides as visible
+			/*
+				Loop through blocks in sector
+			*/
+			for (MapBlock *block : sectorblocks) {
+				MapBlockMesh *mesh = block ? block->mesh : nullptr;
 
-	std::set<v3s16> shortlist;
-	MeshGrid mesh_grid = m_client->getMeshGrid();
+				// Calculate the coordinates for range and frutum culling
+				v3f mesh_sphere_center;
+				f32 mesh_sphere_radius;
 
-	// Recursively walk the space and pick mapblocks for drawing
-	while (blocks_to_consider.size() > 0) {
+				v3s16 block_pos_nodes = block->getPos() * MAP_BLOCKSIZE;
 
-		v3s16 block_coord = blocks_to_consider.front();
-		blocks_to_consider.pop();
+				if (mesh) {
+					mesh_sphere_center = intToFloat(block_pos_nodes, BS)
+							+ mesh->getBoundingSphereCenter();
+					mesh_sphere_radius = mesh->getBoundingRadius();
+				}
+				else {
+					mesh_sphere_center = intToFloat(block_pos_nodes, BS) + v3f((MAP_BLOCKSIZE * 0.5f - 0.5f) * BS);
+					mesh_sphere_radius = 0.0f;
+				}
+				// Frustum culling
+				// Only do coarse culling here, to account for fast camera movement.
+				// This is needed because this function is not called every frame.
+				float frustum_cull_extra_radius = 300.0f;
+				if (is_frustum_culled(mesh_sphere_center,
+						mesh_sphere_radius + frustum_cull_extra_radius))
+					continue;
+
+				// Raytraced occlusion culling - send rays from the camera to the block's corners
+				if (occlusion_culling_enabled && m_enable_raytraced_culling &&
+						block && mesh &&
+						isBlockOccluded(block, cam_pos_nodes)) {
+					blocks_occlusion_culled++;
+					continue;
+				}
 
-		auto &flags = blocks_seen.getChunk(block_coord).getBits(block_coord);
+				if (mesh_grid.cell_size > 1) {
+					// Block meshes are stored in the corner block of a chunk
+					// (where all coordinate are divisible by the chunk size)
+					// Add them to the de-dup set.
+					shortlist.emplace(mesh_grid.getMeshPos(block->getPos()));
+					// All other blocks we can grab and add to the keeplist right away.
+					if (block) {
+						m_keeplist.push_back(block);
+						block->refGrab();
+					}
+				}
+				else if (mesh) {
+					// without mesh chunking we can add the block to the drawlist
+					block->refGrab();
+					m_drawlist.emplace(block->getPos(), block);
+				}
+			}
+			assert(m_drawlist.empty() || shortlist.empty());
+			for (auto pos : shortlist) {
+				MapBlock * block = getBlockNoCreateNoEx(pos);
+				if (block) {
+					block->refGrab();
+					m_drawlist.emplace(pos, block);
+				}
+			}
+		}
+	}
+	else {
 
-		// Only visit each block once (it may have been queued up to three times)
-		if ((flags & 0x80) == 0x80)
-			continue;
-		flags |= 0x80;
+		// Uncomment to debug occluded blocks in the wireframe mode
+		// TODO: Include this as a flag for an extended debugging setting
+		// if (occlusion_culling_enabled && m_control.show_wireframe)
+		// 	occlusion_culling_enabled = porting::getTimeS() & 1;
```

Comment:
also this part shouldn't be inside the if

Reply:
This was added for the new culler, so I left it there.

Reply:
Moved it up.